### PR TITLE
osd: bail out if transaction size overflows

### PR DESCRIPTION
**Pull request by tchaikov**
with a large MOSDMap message, the transaction size could be greater than
UINT_MAX. so fail early with error messages.

Fixes: http://tracker.ceph.com/issues/16982
Signed-off-by: Kefu Chai <kchai@redhat.com>

### CODE REVIEW

**Comment by athanatos** [comment]
Let's talk about this in standup tomorrow.

**Comment by tchaikov** [comment]
will check if we have a guard for file journal because the transaction size could be greater than the size of file journal!

**Comment by tchaikov** [comment]
changelog

- assert() if transaction size is greater than the journal size.

File: src/osd/OSD.cc [hunk 1]
```diff
@@ -6704,10 +6704,16 @@ void OSD::handle_osd_map(MOSDMap *m)
...
 
   ObjectStore::Transaction t;
+  uint64_t txn_size = 0;
 
   // store new maps: queue for disk and put in the osdmap cache
   epoch_t start = MAX(superblock.newest_map + 1, first);
   for (epoch_t e = start; e <= last; e++) {
+    if (txn_size > t.get_num_bytes()) {
+      derr << __func__ << " transaction size overflowed" << dendl;
+      assert(txn_size > t.get_num_bytes());
```
Comment (xiexingguo) [inline src/osd/OSD.cc]: @tchaikov  I think this assert has no effect(always true)?

Comment (tchaikov) [inline src/osd/OSD.cc]: right. will fix.

**Comment by tchaikov** [comment]
changelog

- should assert() if `assert(txn_size < t.get_num_bytes())`
- compare with `header.max_size` instead of `max_size`, the latter is the size of device not the size of journal.

**Comment by athanatos** [comment]
lgtm, adding needs-qa

**Comment by tchaikov** [comment]
tested at http://pulpito.ceph.com/kchai-2016-08-20_18:41:17-rados-wip-kefu-testing---basic-mira/

failed ones are tracked by 

* http://tracker.ceph.com/issues/17087
* http://tracker.ceph.com/issues/16157
* http://tracker.ceph.com/issues/16800

**Pull request merged**